File: backend/app/schemas/afe.py
"""Pydantic schemas for AFE Management (AFE + AFE Cost Estimation)."""

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.master_data import BlankStr, FlagBool
from app.schemas.rig_well import WellConfigurationOut

AfeType = Literal["Drilling", "Completion"]
AfeStatus = Literal["draft", "submitted", "approved"]
ChargingBasis = Literal["Daily Rate", "Per Service Rate", "Per Section Rate"]
QuantityUnit = Literal["days", "hours"]


# ---------------------------------------------------------------------------
# AFE
# ---------------------------------------------------------------------------


class AfeIn(BaseModel):
    """Create / update payload for an AFE header."""

    afe_code: str = Field(..., min_length=1, max_length=50)
    afe_name: str = Field(..., min_length=1, max_length=200)
    afe_type: AfeType = "Drilling"
    rig_id: int = Field(..., description="Rig the AFE belongs to")
    well_id: int = Field(..., description="Well the AFE is scoped to")
    remarks: str | None = None


class AfeUpdate(BaseModel):
    afe_code: str | None = Field(None, min_length=1, max_length=50)
    afe_name: str | None = Field(None, min_length=1, max_length=200)
    afe_type: AfeType | None = None
    rig_id: int | None = None
    well_id: int | None = None
    remarks: str | None = None


class AfeOut(BaseModel):
    id: int
    afe_code: BlankStr = ""
    afe_name: BlankStr = ""
    afe_type: BlankStr = "Drilling"
    rig_id: int
    well_id: int
    remarks: str | None = None
    status: BlankStr = "draft"
    status_remarks: str | None = None
    submitted_at: datetime | None = None
    approved_at: datetime | None = None
    is_deleted: FlagBool = False
    deleted_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    rig_code: str | None = None
    rig_name: str | None = None
    rig_display: str | None = None
    well_code: str | None = None
    well_name: str | None = None
    well_display: str | None = None
    service_count: int = 0
    consumable_count: int = 0
    tangible_count: int = 0
    estimated_total: Decimal = Decimal("0")

    model_config = ConfigDict(from_attributes=True)


class AfeDropdownOut(BaseModel):
    id: int
    afe_code: str
    afe_name: str
    display_name: str

    model_config = ConfigDict(from_attributes=True)


class AfeStatusIn(BaseModel):
    """Status change — the only place an AFE leaves ``draft``."""

    action: Literal["submit", "approve", "reopen"]
    remarks: str | None = None


# ---------------------------------------------------------------------------
# Cost estimation — inputs
# ---------------------------------------------------------------------------


class ServiceRateIn(BaseModel):
    category: str
    unit_rate: Decimal = Decimal("0")


class ServiceChargeLineIn(BaseModel):
    category: str
    quantity: Decimal = Field(Decimal("0"), ge=0)
    quantity_unit: QuantityUnit = "days"


class ServiceSectionRateIn(BaseModel):
    section_id: int
    phase_id: int | None = None
    amount: Decimal = Decimal("0")


class ServiceLineIn(BaseModel):
    service_id: int
    charging_basis: ChargingBasis = "Daily Rate"
    section_id: int | None = None
    phase_id: int | None = None
    per_service_amount: Decimal = Decimal("0")
    effective_date: date | None = None
    remarks: str | None = None
    rates: list[ServiceRateIn] = Field(default_factory=list)
    charge_lines: list[ServiceChargeLineIn] = Field(default_factory=list)
    section_rates: list[ServiceSectionRateIn] = Field(default_factory=list)


class ConsumableLineIn(BaseModel):
    item_kind: Literal["mud_chemical", "drill_bit", "cement_additive", "fuel"] = "mud_chemical"
    item_id: int | None = None
    item_code: str | None = None
    item_name: str | None = None
    quantity: Decimal = Field(Decimal("1"), ge=0)
    captured_rate: Decimal = Decimal("0")
    override_rate: Decimal | None = Field(None, ge=0)
    uom: str | None = None
    currency: str | None = None
    section_id: int | None = None
    phase_id: int | None = None
    remarks: str | None = None


class TangibleLineIn(BaseModel):
    tangible_id: int
    quantity: Decimal = Field(Decimal("1"), ge=0)
    captured_rate: Decimal = Decimal("0")
    override_rate: Decimal | None = Field(None, ge=0)
    uom: str | None = None
    currency: str | None = None
    remarks: str | None = None


class EstimateIn(BaseModel):
    """The whole estimate of one AFE, saved atomically."""

    services: list[ServiceLineIn] = Field(default_factory=list)
    consumables: list[ConsumableLineIn] = Field(default_factory=list)
    tangibles: list[TangibleLineIn] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Cost estimation — computed output
# ---------------------------------------------------------------------------


class CostComponentOut(BaseModel):
    category: str
    description: str
    quantity: Decimal | None = None
    rate: Decimal | None = None
    unit: str | None = None
    amount: Decimal
    section_label: str | None = None
    phase_label: str | None = None


class LineEstimateOut(BaseModel):
    amount: Decimal
    components: list[CostComponentOut] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class GroupSummaryOut(BaseModel):
    group: str
    amount: Decimal
    line_count: int


class SectionRollupOut(BaseModel):
    section_id: int | None = None
    section_label: str
    planned_days: Decimal
    amount: Decimal


class ServiceLineOut(BaseModel):
    id: int
    service_id: int
    service_code: str | None = None
    service_name: str | None = None
    provider_type: str | None = None
    charging_basis: str
    section_id: int | None = None
    phase_id: int | None = None
    per_service_amount: Decimal = Decimal("0")
    effective_date: date | None = None
    remarks: str | None = None
    rates: list[ServiceRateIn] = Field(default_factory=list)
    charge_lines: list[ServiceChargeLineIn] = Field(default_factory=list)
    section_rates: list[ServiceSectionRateIn] = Field(default_factory=list)
    estimate: LineEstimateOut


class ConsumableLineOut(BaseModel):
    id: int
    item_kind: str
    item_id: int
    item_code: str
    item_name: str
    quantity: Decimal
    captured_rate: Decimal
    override_rate: Decimal | None = None
    uom: str | None = None
    currency: str | None = None
    section_id: int | None = None
    phase_id: int | None = None
    remarks: str | None = None
    estimate: LineEstimateOut


class TangibleLineOut(BaseModel):
    id: int
    tangible_id: int
    tangible_code: str | None = None
    tangible_name: str | None = None
    quantity: Decimal
    captured_rate: Decimal
    override_rate: Decimal | None = None
    uom: str | None = None
    currency: str | None = None
    remarks: str | None = None
    estimate: LineEstimateOut


class AfeEstimateOut(BaseModel):
    """Everything the AFE Cost Estimation tab (and its print sheet) needs."""

    afe: AfeOut
    well_configuration: WellConfigurationOut | None = None
    services: list[ServiceLineOut] = Field(default_factory=list)
    consumables: list[ConsumableLineOut] = Field(default_factory=list)
    tangibles: list[TangibleLineOut] = Field(default_factory=list)
    summary: list[GroupSummaryOut] = Field(default_factory=list)
    by_section: list[SectionRollupOut] = Field(default_factory=list)
    grand_total: Decimal = Decimal("0")
    warnings: list[str] = Field(default_factory=list)

File: frontend/types/afe.ts
/** Type contracts for the AFE Management page and its API payloads. */

export type AfeType = 'Drilling' | 'Completion'
export type AfeStatus = 'draft' | 'submitted' | 'approved'
export type ChargingBasis = 'Daily Rate' | 'Per Service Rate' | 'Per Section Rate'
export type QuantityUnit = 'days' | 'hours'

/** The eight constant charge categories, in the order the UI shows them. */
export const CHARGE_CATEGORIES = [
  'Mobilization',
  'Demobilization',
  'Operation',
  'Standby',
  'Personnel-Operation',
  'Personnel-Standby',
  'Fixed Charge',
  'Others',
] as const

/** Charged once per service, never multiplied by days or sections. */
export const ONE_TIME_CATEGORIES = ['Mobilization', 'Demobilization', 'Fixed Charge'] as const

/** Categories a day quantity is entered against. */
export const DAY_BASED_CATEGORIES = [
  'Operation',
  'Standby',
  'Personnel-Operation',
  'Personnel-Standby',
  'Others',
] as const

export const CHARGING_BASES: ChargingBasis[] = ['Daily Rate', 'Per Service Rate', 'Per Section Rate']

export interface AfeRow {
  id: number
  afe_code: string
  afe_name: string
  afe_type: AfeType
  rig_id: number
  well_id: number
  remarks: string | null
  status: AfeStatus
  status_remarks: string | null
  submitted_at: string | null
  approved_at: string | null
  rig_code: string | null
  rig_name: string | null
  rig_display: string | null
  well_code: string | null
  well_name: string | null
  well_display: string | null
  service_count: number
  consumable_count: number
  tangible_count: number
  estimated_total: string | number
  is_deleted?: boolean
  deleted_at?: string | null
  [key: string]: unknown
}

export interface CostComponent {
  category: string
  description: string
  quantity: string | number | null
  rate: string | number | null
  unit: string | null
  amount: string | number
  section_label: string | null
  phase_label: string | null
}

export interface LineEstimate {
  amount: string | number
  components: CostComponent[]
  warnings: string[]
}

export interface ServiceRateRow {
  category: string
  unit_rate: string | number
}

export interface ServiceChargeRow {
  category: string
  quantity: string | number
  quantity_unit: QuantityUnit
}

export interface ServiceSectionRateRow {
  section_id: number
  phase_id: number | null
  amount: string | number
}

export interface ServiceLineRow {
  id: number
  service_id: number
  service_code: string | null
  service_name: string | null
  provider_type: string | null
  charging_basis: ChargingBasis
  section_id: number | null
  phase_id: number | null
  per_service_amount: string | number
  effective_date: string | null
  remarks: string | null
  rates: ServiceRateRow[]
  charge_lines: ServiceChargeRow[]
  section_rates: ServiceSectionRateRow[]
  estimate: LineEstimate
}

export interface ConsumableLineRow {
  id: number
  item_kind: 'mud_chemical' | 'drill_bit'
  item_id: number
  item_code: string
  item_name: string
  quantity: string | number
  captured_rate: string | number
  override_rate: string | number | null
  uom: string | null
  currency: string | null
  section_id: number | null
  phase_id: number | null
  remarks: string | null
  estimate: LineEstimate
}

export interface TangibleLineRow {
  id: number
  tangible_id: number
  tangible_code: string | null
  tangible_name: string | null
  quantity: string | number
  captured_rate: string | number
  override_rate: string | number | null
  uom: string | null
  currency: string | null
  remarks: string | null
  estimate: LineEstimate
}

export interface GroupSummary {
  group: string
  amount: string | number
  line_count: number
}

export interface SectionRollup {
  section_id: number | null
  section_label: string
  planned_days: string | number
  amount: string | number
}

export interface EstimateSection {
  id: number
  section_id: number
  section_code: string | null
  section_name: string | null
  from_depth: string | number
  to_depth: string | number
  remarks: string | null
  total_days: string | number
  phases: {
    id: number
    phase_id: number
    phase_code: string | null
    phase_name: string | null
    days: string | number
    remarks: string | null
  }[]
}

export interface EstimateWellConfiguration {
  well_id: number
  well_code: string
  well_name: string
  rig_code: string | null
  rig_name: string | null
  status: string
  config_status: string
  depth_unit: string
  total_depth: string | number | null
  total_days: string | number
  sections: EstimateSection[]
}

/** The read model behind the AFE Cost Estimation tab and its print sheet. */
export interface AfeEstimate {
  afe: AfeRow
  well_configuration: EstimateWellConfiguration | null
  services: ServiceLineRow[]
  consumables: ConsumableLineRow[]
  tangibles: TangibleLineRow[]
  summary: GroupSummary[]
  by_section: SectionRollup[]
  grand_total: string | number
  warnings: string[]
}

/** Payload accepted by `PUT /afe/estimates/{id}`. */
export interface EstimatePayload {
  [key: string]: unknown
  services: {
    service_id: number
    charging_basis: ChargingBasis
    section_id: number | null
    phase_id: number | null
    per_service_amount: string | number
    effective_date: string | null
    remarks: string | null
    rates: ServiceRateRow[]
    charge_lines: ServiceChargeRow[]
    section_rates: ServiceSectionRateRow[]
  }[]
  consumables: {
    item_kind: 'mud_chemical' | 'drill_bit'
    item_id: number
    quantity: string | number
    captured_rate: string | number
    override_rate: string | number | null
    uom: string | null
    currency: string | null
    section_id: number | null
    phase_id: number | null
    remarks: string | null
  }[]
  tangibles: {
    tangible_id: number
    quantity: string | number
    captured_rate: string | number
    override_rate: string | number | null
    uom: string | null
    currency: string | null
    remarks: string | null
  }[]
}

/** Master-data lookups used by the pickers. */
export interface ServiceOption {
  id: number
  service_code: string
  service_name: string
  provider_type: string
  vendor_display?: string | null
}

export interface ConsumableOption {
  id: number
  code: string
  name: string
  rate: number
  uom: string | null
  currency: string | null
  kind: 'mud_chemical' | 'drill_bit'
  detail: string
}

export interface TangibleOption {
  id: number
  code: string
  name: string
  rate: number
  uom: string | null
  currency: string | null
  detail: string
}
